### DISCUSSION TITLE
TreeSelect component

### DISCUSSION THREAD

**Discussion by wang2lang** [mantinedev/mantine#1051] [Feature requests]
### Usage example, including component/hook API

https://github.com/react-component/tree
https://github.com/react-component/tree-select

**Comment by zhulien-ivanov** [comment]
Those are the only 2 essential components for a UI library that are still missing. We're planning migration to Mantine but can't do it without a TreeSelect of some sort. Do you have any plans on implementing this for the next major release?

**Comment by pavi2410** [comment]
@rtivital Let me know if you require any help. I need this component.

**Reply by pavi2410** [reply]
Update: I used nested `List` component to achieve a tree layout.

```jsx
function TreeList({form, selected, setSelected}) {
  return (
    <List listStyleType="none" withPadding>
      <TreeNode component={form} selected={selected} setSelected={setSelected}/>
    </List>
  )
}

function TreeNode({component, selected, setSelected}) {

  function onSelect(e) {
    e.stopPropagation()
    setSelected(component)
  }

  return (
    <List.Item onClick={onSelect}>
      <Stack style={{backgroundColor: component === selected && "#4dabf733"}}>
        {component.name}
      </Stack>
      <List listStyleType="none" withPadding style={{borderLeft: "2px solid #ddd"}}>
        {
          component.children?.map((child, i) => (
            <TreeNode component={child} key={i} selected={selected} setSelected={setSelected}/>
          ))
        }
      </List>
    </List.Item>
  )
}
```

**Reply by pavi2410** [reply]
https://carbondesignsystem.com/components/tree-view/usage/

This could be a good guideline to begin with

**Reply by pavi2410** [reply]
I was able to make this component using just the List component and nest it.

![image](https://github.com/mantinedev/mantine/assets/28837746/c5f80d1c-2c5a-4e7b-99a4-69a1c9101ca8)

**Reply by stephenirven** [reply]
That looks pretty good. Does it support multi/select also? 

Either way if you'd be prepared to share the code for that, it might be useful to form the basis of a new component.

**Reply by pavi2410** [reply]
Here's the component I built: https://github.com/Kodular/ai-unchive/blob/d75102e2d534b72609a832da5218250a5c83a87d/src/components/LayoutPanel.tsx

It doesn't support multi-select or collapsing children nodes.

Anyway, I am in the process of migrating to v7, which breaks a lot of things for me. I'd contribute this component after I am done with the migration.

**Reply by pavi2410** [reply]
I just migrated this component to v7 and added collapsibility.

![image](https://github.com/mantinedev/mantine/assets/28837746/d2977fe9-2ca7-46a2-899f-95c5f08dbc88)

**Reply by pavi2410** [reply]
This might look better now. I changed how indents are positioned.
![image](https://github.com/mantinedev/mantine/assets/28837746/f76aa290-c28c-4e34-9ce1-e010e86ea60e)

**Reply by pavi2410** [reply]
This is definitely a lot inspired by VS Code.

**Reply by pavi2410** [reply]
Where was this hiding for so long?
 https://www.w3.org/WAI/ARIA/apg/patterns/treeview/

**Reply by pavi2410** [reply]
I've moved to a different UI library which has more enterprise components such as the Tree. See https://primereact.org

**Comment by stephenirven** [comment]
I'm also interested in this. For info of anyone looking for a solution right now, it could be worth looking at this [https://github.com/minop1205/react-dnd-treeview](https://github.com/minop1205/react-dnd-treeview)

It seems to be quite fully featured, and there are CodeSandbox samples for various use cases including select. The samples use Material-UI, but the module doesn't depend on it, so it's possible to get it working with Mantine.

**Reply by lifeparticle** [reply]
A `treeview` would be a great component. Thanks!

**Comment by adamerose** [comment]


**Reply by rtivital** [reply]
Issues are now used to track bugs, feature requests are just proposals that we can consider implementing, while are the real issues we are planning to resolve.

**Comment by iakovalgaev** [comment]
Yes, that would be cool ! I'm waiting =)

There are already groups now, it would be great if the list could collapse.

**Comment by zhulien-ivanov** [comment]
Is this still planned for an upcoming release (including v5)?

**Reply by rtivital** [reply]
No, it is not planned as a part of 5.0 release, I doubt that adding this component worth the effort – I've never seen it being used in real applications

**Reply by zhulien-ivanov** [reply]
It's widely used in business applications. It basically covers any case where you want to present options in a categorized way. Talking about TreeSelect-like component.

**Reply by pavi2410** [reply]
There seems to be many kinds of treeview select in UI design, such inside a dropdown, toolbar, or as a list with/without collapsible variants.

It's too broad to implement. I think the better way to implement this is to extend existing components.

Here's a mapping of components that can be extended:
Select -> TreeSelect
Menu -> Nested Menu
List -> TreeList (with selectable and collapsible nodes)

**Reply by zhulien-ivanov** [reply]
We're talking specifically about https://github.com/react-component/tree-select. It's basically like a `MultiSelect` but supports nested items.

**Reply by rtivital** [reply]
I'm planning to add Tree component, we'll see how it goes from there

**Reply by elarens** [reply]
@rtivital Is there any timeline for that? I need to use a tree component for my dropdown menu but can't find anything headless and well maintained that I could easily adapt :(((

**Reply by iakovalgaev** [reply]
I also support the posts above !

**Comment by suchintan** [comment]
I'm looking forward to this feature as well!

**Comment by suchintan** [comment]
FYI I ended up building my own using react-dnd-treeview. It ended up playing well with mantine

**Reply by husniramdani** [reply]
this one https://www.npmjs.com/package/@minoru/react-dnd-treeview?

**Comment by DivMode** [comment]
Also looking for one with checkboxes, something like this from antd: https://codesandbox.io/s/h3ny7f?file=/index.tsx

**Comment by carafry** [comment]
Is this still on the roadmap?  

It would be really useful to have a checkable tree select component similar to Ant Design: https://ant.design/components/tree-select

<img width="586" alt="Screenshot 2023-03-23 at 2 41 42 PM" src="https://user-images.githubusercontent.com/104099317/227333520-68e96c73-2601-4af7-9e5e-d3b105d22304.png">

**Comment by langziyang** [comment]
please integrated https://www.npmjs.com/package/rc-tree-select

**Comment by starkblaze01** [comment]
We are doing an evaluation on migrating our UI components, so far Mantine is our most preferred choice, but a direct implementation of tree select would be much appreciated!! Hopefully we will get to see this component to be released soon.

**Reply by starkblaze01** [reply]
Thank you for this: https://mantine.dev/core/tree/ --> v7.10.0+

**Comment by BangHuynh0404** [comment]
This is my custom component, hop this help: 
import { Combobox, useCombobox, Input, Pill, PillsInput, Checkbox } from '@mantine/core'
import { useMemo, useState } from 'react'
import { IconChevronRight, IconChevronDown } from '@tabler/icons-react'

interface TreeItem {
  ID: string
  Term: string
  Children?: TreeItem[]
}

interface TreeMultiSelectProps {
  data: TreeItem[]
  value: string[]
  onChange: (value: string[]) => void
  placeholder?: string
  className?: string
  defaultExpanded?: boolean
}

export function TreeMultiSelect({
  data,
  value,
  onChange,
  placeholder,
  className,
  defaultExpanded = false,
}: TreeMultiSelectProps) {
  const combobox = useCombobox({
    onDropdownClose: () => combobox.resetSelectedOption(),
  })

  const [search, setSearch] = useState('')
  const [expandedNodes, setExpandedNodes] = useState<Set<string>>(() => {
    if (!defaultExpanded) return new Set()

    const expanded = new Set<string>()
    const collectNodes = (items: TreeItem[]) => {
      items.forEach((item) => {
        if (item.Children?.length) {
          expanded.add(item.ID)
          collectNodes(item.Children)
        }
      })
    }
    collectNodes(data)
    return expanded
  })

  const flattenedOptions = useMemo(() => {
    const options: { value: string; label: string; level: number; hasChildren: boolean }[] = []

    function processItem(item: TreeItem, level: number = 0) {
      options.push({
        value: item.ID,
        label: item.Term,
        level,
        hasChildren: !!(item.Children && item.Children.length > 0),
      })

      if (item.Children && item.Children.length > 0 && expandedNodes.has(item.ID)) {
        item.Children.forEach((child) => processItem(child, level + 1))
      }
    }

    data.forEach((item) => processItem(item))
    return options
  }, [data, expandedNodes])

  const handleValueSelect = (val: string) => {
    const newValue = value.includes(val) ? value.filter((v) => v !== val) : [...value, val]
    onChange(newValue)
  }

  const handleValueRemove = (val: string) => {
    onChange(value.filter((v) => v !== val))
  }

  const toggleNode = (nodeId: string) => {
    const newExpanded = new Set(expandedNodes)
    if (newExpanded.has(nodeId)) {
      newExpanded.delete(nodeId)
    } else {
      newExpanded.add(nodeId)
    }
    setExpandedNodes(newExpanded)
  }

  const values = value.map((val) => {
    const option = flattenedOptions.find((opt) => opt.value === val)
    return (
      <Pill key={val} withRemoveButton onRemove={() => handleValueRemove(val)}>
        {option?.label || val}
      </Pill>
    )
  })

  const filteredOptions = flattenedOptions.filter((option) =>
    option.label.toLowerCase().includes(search.toLowerCase())
  )

  return (
    <Combobox store={combobox} onOptionSubmit={handleValueSelect}>
      <Combobox.Target>
        <PillsInput className={className} onClick={() => combobox.openDropdown()}>
          <Pill.Group>
            {values}
            <Combobox.EventsTarget>
              <PillsInput.Field
                placeholder={placeholder || 'Select items'}
                value={search}
                onChange={(e) => setSearch(e.currentTarget.value)}
                onFocus={() => combobox.openDropdown()}
                onBlur={() => combobox.closeDropdown()}
                onKeyDown={(event) => {
                  if (event.key === 'Backspace' && search.length === 0) {
                    event.preventDefault()
                    handleValueRemove(value[value.length - 1])
                  }
                }}
              />
            </Combobox.EventsTarget>
          </Pill.Group>
        </PillsInput>
      </Combobox.Target>

      <Combobox.Dropdown>
        <Combobox.Options style={{ maxHeight: '300px', overflowY: 'auto' }}>
          {filteredOptions.length > 0 ? (
            filteredOptions.map((option) => (
              <Combobox.Option
                key={option.value}
                value={option.value}
                style={{
                  paddingLeft: `${option.level * 20 + 12}px`,
                  minHeight: '36px',
                  display: 'flex',
                  alignItems: 'center',
                }}
              >
                <div className="flex items-center gap-2">
                  {option.hasChildren && (
                    <div
                      onClick={(e) => {
                        e.stopPropagation()
                        toggleNode(option.value)
                      }}
                      className="cursor-pointer"
                    >
                      {expandedNodes.has(option.value) ? (
                        <IconChevronDown size={16} />
                      ) : (
                        <IconChevronRight size={16} />
                      )}
                    </div>
                  )}
                  {!option.hasChildren && <div style={{ width: 16 }} />}
                  <Checkbox
                    checked={value.includes(option.value)}
                    onChange={(e) => {
                      e.stopPropagation()
                      handleValueSelect(option.value)
                    }}
                    onClick={(e) => e.stopPropagation()}
                  />
                  {option.label}
                </div>
              </Combobox.Option>
            ))
          ) : (
            <Combobox.Empty>Nothing found</Combobox.Empty>
          )}
        </Combobox.Options>
      </Combobox.Dropdown>
    </Combobox>
  )
}

**Comment by nepton-mattpike** [comment]
We would like to use an official tree select component please. Also is it easy to add lazy loading to the tree components? an example would be great, we have large trees
